enhancements/cvo/cluster-profiles.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Cluster Profiles
+authors:
+  - "@csrwng"
+reviewers:
+  - "@abhinavdahiya"
+  - "@crawford"
+  - "@sdodson"
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+approvers:
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+creation-date: 2020-02-04
+last-updated: 2019-02-04
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# cluster-profiles
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Cluster profiles are a way to support different deployment models for OpenShift clusters. 
+A profile is an identifier that the Cluster Version Operator uses to determine
+which manifests to apply. Operators can be excluded completely or can have different
+manifests for each supported profile.
+
+## Motivation
+
+In order to support different deployment models in which not all operators rendered by
+the CVO by default are needed, a mechanism is needed by which we tell the CVO which 
+manifests to include or exclude. One such deployment is the hosted control plane case
+in which the control plane components and corresponding operators should not be rendered
+by the CVO.
+
+
+### Goals
+
+- Equip the CVO to include/exclude manifests based on a matching identifier (profile)
+
+### Non-Goals
+
+- Define which manifests should be excluded/included for each use case
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+As a user, I can create a cluster in which manifests for control plane operators are
+not applied by the CVO.
+
+#### Story 2
+As a user, I can create a cluster in which node selectors for certain operators target
+worker nodes instead of master nodes.
+
+### Design
+
+A cluster profile is specified to the CVO as an identifier in an environment
+variable. For a given cluster, only one CVO profile may be in effect.
+
+```
+CLUSTER_PROFILE=[identifier]
+```
+This environment variable would have to be specified in the CVO deployment. When
+no `CLUSTER_PROFILE=[identifier]` variable is specified, the default cluster profile
+is in effect.
+
+The following annotations may be used to include/exclude manifests for a given profile:
+
+```
+exclude.release.openshift.io/[identifier]=true
```

Comment:
I think we should avoid excludes.  Every labels selector we have today is a positive selection.  Negative exclusion will include manifests without the owners of those manifests choosing to support a profile.  Instead of owners knowing the profiles they need to support, we'll end up with case where bugs are opened against components who then respond with, "wait, what profile?  that isn't going to work".

To the common counter argument of "but then it takes a lot of work to create a new profile and all the teams have to aware of it", I say "**good!**".  We don't want that many profiles or our support matrix will grow and customers will be trying to figure out which profile is right for them.  Profiles should come with a cost and if component owners are expected to support their components in these profiles, they really should be aware even if it's exactly an optional assignment.

Reply:
That said, an alternative would be to have a positive list somewhere else which qualifies an operator for a certain profile. Then excludes might make it easier to express profiles.

Reply:
@deads2k i think this is a good argument.